### PR TITLE
Show quick pick when redirect fails to complete redirect to be more robust

### PR DESCRIPTION
**Pull request by TylerLeonhardt**
We were seeing some folks who are running code-server on a different url (like an ip address of an Azure VM) were struggling to complete the auth flow.

To assist, this shows an input box now where the user can paste in their authorization code to complete the auth flow.

image

### CODE REVIEW

File: extensions/microsoft-authentication/src/utils.ts [hunk 1]
```diff
@@ -2,7 +2,31 @@
...
 export function toBase64UrlEncoding(base64string: string) {
 	return base64string.replace(/=/g, '').replace(/\+/g, '-').replace(/\//g, '_'); // Need to use base64url encoding
 }
+
+export function isSupportedEnvironment(uri: Uri): boolean {
+	if (env.uiKind === UIKind.Desktop) {
+		return true;
+	}
+	// local development (localhost:* or 127.0.0.1:*)
+	if (/^https?$/i.test(uri.scheme) && /^(localhost|127\.0\.0\.1):\d+$/i.test(uri.authority)) {
```
Comment (mjbvz) [inline extensions/microsoft-authentication/src/utils.ts:16]: you may want to use `isLocalHost` for this: https://github.com/microsoft/vscode/blob/8fa5722aad72c6712c821bc1deeede56966b5d84/src/vs/platform/tunnel/common/tunnel.ts#L149

File: extensions/microsoft-authentication/src/AADHelper.ts [hunk 2]
```diff
@@ -680,8 +675,32 @@ export class AzureActiveDirectoryService {
...
+			inputBox.onDidAccept(async () => {
+				const code = inputBox.value;
+				if (code) {
+					const session = await this.exchangeCodeForSession(code, verifier, scopeData);
+					resolve(session);
+				}
+			});
+			inputBox.onDidHide(() => {
+				if (!inputBox.value) {
+					inputBox.dispose();
```
Comment (mjbvz) [inline extensions/microsoft-authentication/src/AADHelper.ts:694]: Should the input box also be disposed when the user accepts something?

**Review by mjbvz — DISMISSED** [review]

File: extensions/microsoft-authentication/src/AADHelper.ts (hunk 2 above)
Comment (TylerLeonhardt) [reply extensions/microsoft-authentication/src/AADHelper.ts:695]: Nice catch. Fixed.

File: extensions/microsoft-authentication/src/utils.ts (hunk 1 above)
Comment (TylerLeonhardt) [reply extensions/microsoft-authentication/src/utils.ts]: switched to using that. (this is in extension land so I can't leverage that directly)

**Review by mjbvz — APPROVED** [review]

**Pull request merged**